---
title: "Comparing DHS, H3K27ac regions and cell type specific peaks from ATAC-seq"
author: "XSun"
date: "2022-09-14"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

```{r, echo = FALSE,message=FALSE, warning=FALSE}
suppressMessages(library(ggplot2))
suppressMessages(library(GenomicRanges))
suppressMessages(library(ComplexHeatmap))
suppressMessages(library(genomation))


load("/project2/xinhe/xsun/heart_atlas/5.OR/data/OCR_celltypespecific.rdata")
load("/project2/xinhe/xsun/heart_atlas/5.OR/data/encode.beds.rdata")


subsetByOverlapProp <- function(q, s, minPoverlap, maxgap=0){
  
  hits <- GenomicRanges::findOverlaps(query = q, subject = s, maxgap = maxgap)
  overlaps <- pintersect(q[queryHits(hits)], s[subjectHits(hits)])
  percentOverlap <- width(overlaps) / width(q[queryHits(hits)])
  hits <- hits[percentOverlap >= minPoverlap]
  
  return(hits)
}

ggClean <- function(rotate_axis=FALSE){
  tm <- theme_bw() + 
    theme(text = element_text(size=18),
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_rect(colour = "black", size=1))
  if(rotate_axis){
    tm <- tm + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  }
  
  return(tm)
  
}


```


## Introduction

We will try to address the 3rd comments from the 3rd reviewer here: 

Given that the numbers of cell type-specific peaks across different cell types are not the same, it does not appear to be valid to just show the proportions of overlap with bulk level data (Fig. 3d). The odd ratios are useful to correct differences in peak numbers. However, it would be helpful to show how many informative peaks are exclusively identified by single-cell data. 

We computed odd ratios according to:

$$ OR = \frac{\frac{P(OCR_{2}|OCR_{1})}{1-P(OCR_{2}|OCR_{1})}}{\frac{P(OCR_{2}|\overline{OCR_{1}})}{1-P(OCR_{2}|\overline{OCR_{1}})}} $$
where $OCR_{1}$ is the bulk peaks from DHS, H3K27ac regions and $OCR_{2}$ are peaks for each cell type specific OCR peaks. We permuted $OCR_{1}$ among whole genome (hg38,$OCR_{1}$ themselves were excluded) and used these regions as $\overline{OCR_{1}}$ . 

The permutation were performed using bedtools under two settings: 1) permuted 10 times the $OCR_{1}$ 2) permuted same size of $OCR_{1}$.

The H3K27ac regions covers 1/16 of the whole genome, bedtools cannot handle permutation for the data 10 times of it. So the permutation for H3K27ac region was just performed under the second setting. 

For DHS peaks, the OR results were not differ a lot under different settings. 

# Results for 10 times permutation 


```{r echo=F, message=FALSE, warning=FALSE, paged.print=FALSE,fig.height=5, fig.width=6}

odd_all <- c()
for (i in 1:length(encode.beds)) {
  
  gr_random <- readBed(paste0("/project2/xinhe/xsun/heart_atlas/5.OR/data/data_shuffled_exclude_itself/",names(encode.beds)[i],"_shuffled.bed"))
  gr_bulk <- encode.beds[[i]]
  
  odd1 <- c()
  odd2 <- c()
  for (j in 1:length(markers)) {
    
    gr_celltype <- markers[[j]]
    
    hit_overlap_with_bulk <- subsetByOverlapProp(q = gr_celltype, s = gr_bulk, minPoverlap = 0.1)
    overlap_with_bulk <- gr_celltype[unique(queryHits(hit_overlap_with_bulk)),]
    
    hit_overlap_with_random <- subsetByOverlapProp(q = gr_celltype, s = gr_random, minPoverlap = 0.1)
    overlap_with_random <- gr_celltype[unique(queryHits(hit_overlap_with_random)),]
    
    odd1[j] <- (length(overlap_with_bulk) / length(gr_bulk)) / ( 1- (length(overlap_with_bulk) / length(gr_bulk)))
    odd2[j] <- (length(overlap_with_random) / length(gr_random))  / (1 - (length(overlap_with_random) / length(gr_random))) 
    
  }
  odd <- odd1/odd2
  odd_all <- rbind(odd_all,odd)
  
}

colnames(odd_all) <- names(markers)
rownames(odd_all) <- names(encode.beds)

odd_re <- odd_all[,c("Cardiomyocyte","Pericyte","Endothelial","Fibroblast","Lymphoid","Myeloid")]
odd_heatmap <- odd_re[c("Heart_RVLV","DermisEndothelialBlood","Tcell","Bcell","CD14Monocyte"),]

rownames(odd_heatmap) <- c("Heart \nLV+RV","Dermis \nEndothelial","T cell","B cell","CD14+ \nMonocyte")
colnames(odd_heatmap) <- c("CM","Peri","Endo","Fibro","Lymph", "Mye")

DT::datatable(odd_heatmap, options = list(pageLength =10))

Heatmap(matrix = odd_heatmap, cluster_rows = F, 
        cluster_columns = F, show_row_names = T, row_names_side = "left",
        name = "Odd Ratio",
        col = circlize::colorRamp2(c(0,9,62), c("lightblue","white","firebrick")), 
        row_title = "ENCODE DNase",
        column_title = "Cell Types",
        row_gap = unit(1, "mm"),
        column_gap = unit(1, "mm"),
        na_col = "white",
        rect_gp = gpar(col = "black", lwd = 0.5),
        use_raster = F)


odd_log <- log(odd_heatmap)
Heatmap(matrix = odd_log, cluster_rows = F, 
        cluster_columns = F, show_row_names = T, row_names_side = "left",
        name = "log(OR)",
        col = circlize::colorRamp2(c(0,2,4), c("lightblue","white","firebrick")), 
        row_title = "ENCODE DNase",
        column_title = "Cell Types",
        row_gap = unit(1, "mm"),
        column_gap = unit(1, "mm"),
        na_col = "white",
        rect_gp = gpar(col = "black", lwd = 0.5),
        use_raster = F)
```



# Results for same size permutation

```{r echo=F, message=FALSE, warning=FALSE, paged.print=FALSE,fig.height=5, fig.width=6}

odd_all <- c()
for (i in 1:length(encode.beds)) {
  
  gr_random <- readBed(paste0("/project2/xinhe/xsun/heart_atlas/5.OR/data/data_shuffled_exclude_itself_samesize//",names(encode.beds)[i],"_shuffled.bed"))
  gr_bulk <- encode.beds[[i]]
  
  odd1 <- c()
  odd2 <- c()
  for (j in 1:length(markers)) {
    
    gr_celltype <- markers[[j]]
    
    hit_overlap_with_bulk <- subsetByOverlapProp(q = gr_celltype, s = gr_bulk, minPoverlap = 0.1)
    overlap_with_bulk <- gr_celltype[unique(queryHits(hit_overlap_with_bulk)),]
    
    hit_overlap_with_random <- subsetByOverlapProp(q = gr_celltype, s = gr_random, minPoverlap = 0.1)
    overlap_with_random <- gr_celltype[unique(queryHits(hit_overlap_with_random)),]
    
    odd1[j] <- (length(overlap_with_bulk) / length(gr_bulk)) / ( 1- (length(overlap_with_bulk) / length(gr_bulk)))
    odd2[j] <- (length(overlap_with_random) / length(gr_random))  / (1 - (length(overlap_with_random) / length(gr_random))) 
    
  }
  odd <- odd1/odd2
  odd_all <- rbind(odd_all,odd)
  
}

colnames(odd_all) <- names(markers)
rownames(odd_all) <- names(encode.beds)

odd_re <- odd_all[,c("Cardiomyocyte","Pericyte","Endothelial","Fibroblast","Lymphoid","Myeloid")]
odd_heatmap <- odd_re[c("Heart_RVLV","DermisEndothelialBlood","Tcell","Bcell","CD14Monocyte"),]

rownames(odd_heatmap) <- c("Heart \nLV+RV","Dermis \nEndothelial","T cell","B cell","CD14+ \nMonocyte")
colnames(odd_heatmap) <- c("CM","Peri","Endo","Fibro","Lymph", "Mye")

DT::datatable(odd_heatmap, options = list(pageLength =10))

Heatmap(matrix = odd_heatmap, cluster_rows = F, 
        cluster_columns = F, show_row_names = T, row_names_side = "left",
        name = "Odd Ratio",
        col = circlize::colorRamp2(c(0,9,62), c("lightblue","white","firebrick")), 
        row_title = "ENCODE \nDNase",
        column_title = "Cell Types",
        row_gap = unit(1, "mm"),
        column_gap = unit(1, "mm"),
        na_col = "white",
        rect_gp = gpar(col = "black", lwd = 0.5),
        use_raster = F)


odd_log <- log(odd_heatmap)
Heatmap(matrix = odd_log, cluster_rows = F, 
        cluster_columns = F, show_row_names = T, row_names_side = "left",
        name = "log(OR)",
        col = circlize::colorRamp2(c(0,2,4), c("lightblue","white","firebrick")), 
        row_title = "ENCODE \nDNase",
        column_title = "Cell Types",
        row_gap = unit(1, "mm"),
        column_gap = unit(1, "mm"),
        na_col = "white",
        rect_gp = gpar(col = "black", lwd = 0.5),
        use_raster = F)
```

```{r echo=F, message=FALSE, warning=FALSE, paged.print=FALSE,fig.height=5, fig.width=5}
odd_h3k <- data.frame(odd_re["Heart_LVRV",])
odd_h3k <- cbind(rownames(odd_h3k),odd_h3k)
colnames(odd_h3k) <- c("celltypes","Odd_ratio")
odd_h3k$celltypes <- c("CM","Peri","Endo","Fibro","Lymph", "Mye")
odd_h3k$celltypes <- factor(odd_h3k$celltypes,levels = odd_h3k$celltypes)


ggplot(odd_h3k, aes(x=celltypes, y=Odd_ratio)) + 
  geom_bar(stat='identity', width=0.7) + 
  ggClean(rotate_axis = T) + 
  ylab('Odd Ratio') + 
  xlab('')

odd_h3k$Odd_ratio <- log(odd_h3k$Odd_ratio)
ggplot(odd_h3k, aes(x=celltypes, y=Odd_ratio)) + 
  geom_bar(stat='identity', width=0.7) + 
  ggClean(rotate_axis = T) + 
  ylab('log(OR)') + 
  xlab('')



```
